### PR TITLE
Replace `structurally_resolved_type` in casts check.

### PR DESCRIPTION
**Pull request by leoyvens**
The behaviour of `resolve_type_vars_if_possible` is simpler and infallible. Other minor refactorings.

I'm not sure if this is backwards compatible, in theory resolving obligations between two cast checks could solve a dependency between them, but I don't know if that's actually possible and it doesn't sound like something we'd want to support.

### CODE REVIEW

**Comment by estebank** [comment]
LGTM, I think we should run crater on this to verify that we're not breaking something in the wild.

@bors try

**Comment by leodasvacas** [comment]
I've tacked on a similar refactoring to unary expr check. In theory this could make more code compile because we don't error or could ommit an otherwise useful error. In practice I couldn't observe any effect.

**Comment by estebank** [comment]
cc @rust-lang/compiler

File: src/librustc/ty/cast.rs [hunk 1]
```diff
@@ -64,7 +63,7 @@ impl<'tcx> CastTy<'tcx> {
             ty::TyBool => Some(CastTy::Int(IntTy::Bool)),
             ty::TyChar => Some(CastTy::Int(IntTy::Char)),
             ty::TyInt(_) => Some(CastTy::Int(IntTy::I)),
-            ty::TyInfer(ty::InferTy::IntVar(_)) => Some(CastTy::Int(IntTy::Ivar)),
+            ty::TyInfer(ty::InferTy::IntVar(_)) => Some(CastTy::Int(IntTy::I)),
```
Comment (nikomatsakis) [inline src/librustc/ty/cast.rs]: this `Ivar` is basically unused I guess?

File: src/librustc_typeck/check/cast.rs [hunk 2]
```diff
@@ -488,11 +488,7 @@ impl<'a, 'gcx, 'tcx> CastCheck<'tcx> {
                     ty::TypeVariants::TyInfer(t) => {
                         match t {
                             ty::InferTy::IntVar(_) |
-                            ty::InferTy::FloatVar(_) |
-                            ty::InferTy::FreshIntTy(_) |
-                            ty::InferTy::FreshFloatTy(_) => {
```
Comment (nikomatsakis) [inline src/librustc_typeck/check/cast.rs]: Similarly, the "fresh" variants here are just dead code I suppose

**Comment by nikomatsakis** [comment]
OK, well, the change basically makes sense, not a big diff obviously. crater run seems like a fine thing.

cc @rust-lang/infra

**Comment by nikomatsakis** [comment]
> I'm not sure if this is backwards compatible, in theory resolving obligations between two cast checks could solve a dependency between them, but I don't know if that's actually possible and it doesn't sound like something we'd want to support.

I'm not entirely happy with how casts fit into the pipeline, but I suppose we would probably prefer for them to apply "atomically" (maybe?). At least, I don't love the idea of the ordering being overly significant, although I guess it's not the end of the world if it is, since we already have some ordering dependencies from coercions.

**Comment by nikomatsakis** [comment]
cc @rust-lang/infra -- or should it be @rust-lang/release ? -- any word on crater run?

**Comment by Mark-Simulacrum** [comment]
Crater run started; results expected in ~5 days. Our crater queue is quite full right now, so PRs are waiting for a week or two generally to get their turn at crater.

**Comment by aidanhs** [comment]
Crater results: http://cargobomb-reports.s3.amazonaws.com/pr-48270/index.html. 'Blacklisted' crates (spurious failures etc) can be found [here](https://github.com/rust-lang-nursery/crater/blob/master/blacklist.md). If you see any spurious failures not on the list, please make a PR against that file.

(looks like this broke nix and so broke half the world :smile:)

**Comment by leodasvacas** [comment]
Thanks for the crater run! Interesting results. Curiously nix itself didn't regress because the newest version isn't affected. Anyways, the issue is that it's possible to depend on the order in which casts are checked, this compiles today:
```rust
fn main() {
    let x = &"hello";
    let mut y = x as *const _;
    y = 0 as *const _;
}
```
But this does not compile today:
```rust
fn main() {
    let x = &"hello";
    let mut y = 0 as *const _;
    y = x as *const _;
              ^^^^^^^^ cannot cast to a pointer of an unknown kind.
}
```
We should make a decision, either they should both compile, or none of them should compile. If they both should compile, then we should use a more robust fixed-point algorithm rather than check in arbitratry order. If none of them should compile, then we should introduce a future-compatibility warning.

**Comment by nikomatsakis** [comment]
> We should make a decision, either they should both compile, or none of them should compile

Ideally yes, but in fact order does matter to rustc type checking quite a lot -- all coercions operate on a "greedy basis" and it's going to be a long road getting away from it, if we ever fully do.

**Comment by nikomatsakis** [comment]
So, like @leodasvacas, I am not thrilled with the order dependency on casts -- we had intended to avoid it. Though, iirc, I've noted it before, and decided that -- since it exists and code relies on it -- we ought to keep that code working as best we can. Perhaps the same rule of thumb applies here.

I don't think that breaking the older version of nix is really an option. It *might* be ok if we can instead publish a patch for that older version. That's imperfect, since apps would still need to update their lock files, but better than the current status. We might then want to do a warning period as well.

**Comment by nikomatsakis** [comment]
I'd like to nominate for discussion but I'm not sure where =) I guess we'll start with T-lang.

**Comment by nikomatsakis** [comment]
@leodasvacas 

> We should make a decision, either they should both compile, or none of them should compile. If they both should compile, then we should use a more robust fixed-point algorithm rather than check in arbitratry order.

So we were talking about this in the meeting. I find this idea appealing -- but I am a bit wary of building that on the current implementation. I wonder if you'd have interest in trying to extend Chalk to model coercions and casts? In such a context, I would feel much more confident that we aren't accidentally introducing ambiguity. 

In general, I have this feeling like the right way to correct things here is to try and leave the current implementation alone -- so as not to disturb existing code -- and transition towards a new implementation with a cleaner foundation.

**Comment by leodasvacas** [comment]
I've updated this PR to revert the problematic change and added tests for the current behavior.

> I wonder if you'd have interest in trying to extend Chalk to model coercions and casts?

Sounds interesting, I'll ping you on gitter.

**Comment by shepmaster** [comment]
What is the path forward with this PR? It sounds like "close" to me.

**Comment by shepmaster** [comment]
> should be waiting-for-review.

That's on you @nikomatsakis !

**Pull request merged**